util/expensivequery/expensivequery.go

```diff
@@ -73,12 +83,210 @@ func (eqh *Handle) Run() {
 				}
 			}
 			threshold = atomic.LoadUint64(&variable.ExpensiveQueryTimeThreshold)
+			if record.err == nil {
+				record.alarm4ExcessiveMemUsage(sm)
+			}
 		case <-eqh.exitCh:
 			return
 		}
 	}
 }
 
+type memoryUsageAlarm struct {
+	err                          error
+	isServerMemoryQuotaSetByUser bool
+	serverMemoryQuota            uint64
+	lastChangeTime               time.Time
+
+	tmpDir              string
+	lastLogFileName     []string
+	lastProfileFileName [][]string // heap, goroutine
+}
+
+var record memoryUsageAlarm
+
+func initMemoryUsageAlarmRecord() {
+	if alert := config.GetGlobalConfig().Performance.MemoryUsageAlarmRatio; alert == 0 || alert == 1 {
+		record.err = errors.New("close memory usage alarm recorder")
+		return
+	}
+	if quota := config.GetGlobalConfig().Performance.ServerMemoryQuota; quota != 0 {
+		record.serverMemoryQuota = quota
+		record.isServerMemoryQuotaSetByUser = true
+	} else {
+		record.serverMemoryQuota, record.err = memory.MemTotal()
+		if record.err != nil {
+			logutil.BgLogger().Error("get system total memory fail", zap.Error(record.err))
+			return
+		}
```

Comment:
Can we add a TODO here? I think we can investigate how to directly get the memory info in container later.

Reply:
Done